api/paidAction/index.js

```diff
@@ -95,271 +128,285 @@ export default async function performPaidAction (actionType, args, context) {
+      result: await performAction(invoiceEntry, paidAction, args, context),
+      paymentMethod: 'OPTIMISTIC',
+      retriable: invoiceEntry.retriable,
+      retriableWithFeeCredits: invoiceEntry.retriableWithFeeCredits,
+      retriableManually: invoiceEntry.retriableManually
     }
-  }, { isolationLevel: Prisma.TransactionIsolationLevel.ReadCommitted })
-}
-
-async function performPessimisticAction (actionType, args, context) {
-  const action = paidActions[actionType]
-
-  if (!action.supportsPessimism) {
-    throw new Error(`This action ${actionType} does not support pessimistic invoicing`)
   }
+  if (context.tx) return await performInvoicedAction(context.tx)
+  return await models.$transaction(performInvoicedAction, { isolationLevel: Prisma.TransactionIsolationLevel.ReadCommitted })
+}
 
-  // just create the invoice and complete action when it's paid
-  const invoiceArgs = await createLightningInvoice(actionType, args, context)
-  return {
-    invoice: await createDbInvoice(actionType, args, context, invoiceArgs),
-    paymentMethod: 'PESSIMISTIC'
+async function performPessimisticAction (invoiceData, actionType, paidAction, args, context) {
+  const { models, actionId, me } = context
+  const performInvoicedAction = async tx => {
+    context.tx = tx
+    context.optimistic = false
+    await checkUser(me?.id ?? USER_ID.anon, context)
+
+    const invoiceEntry = await createDbInvoice(invoiceData, {
+      actionType,
+      actionId,
+      optimistic: false,
+      args
+    }, context)
+    return {
+      invoice: invoiceEntry,
+      paymentMethod: 'PESSIMISTIC',
+      retriable: invoiceEntry.retriable,
+      retriableWithFeeCredits: invoiceEntry.retriableWithFeeCredits,
+      retriableManually: invoiceEntry.retriableManually
+    }
   }
+  if (context.tx) return await performInvoicedAction(context.tx)
+  return await models.$transaction(performInvoicedAction, { isolationLevel: Prisma.TransactionIsolationLevel.ReadCommitted })
 }
 
-export async function retryPaidAction (actionType, args, context) {
-  const { models, me } = context
-  const { invoice: failedInvoice } = args
+export async function retryPaidAction (actionType, { invoice, forceFeeCredits, forceRetry }, context) {
+  const { models } = context
+  const failedInvoice = invoice
 
-  console.log('retryPaidAction', actionType, args)
+  console.log('retryPaidAction', actionType, invoice)
 
-  const action = paidActions[actionType]
-  if (!action) {
+  const paidAction = paidActions[actionType]
+  if (!paidAction) {
     throw new Error(`retryPaidAction - invalid action type ${actionType}`)
   }
 
-  if (!me) {
-    throw new Error(`retryPaidAction - must be logged in ${actionType}`)
+  if (!failedInvoice) {
+    throw new Error(`retryPaidAction - missing invoice ${actionType}`)
   }
 
-  if (!action.supportsOptimism) {
-    throw new Error(`retryPaidAction - action does not support optimism ${actionType}`)
+  let retriable = false
+  if (forceRetry) {
```

Comment:
the code limits retries so that after a while we know that everything we tried failed, and we consider the action dead (both client and serverside).
However only the user knows if maybe it is a temporary issue with its node or something like that, so there is this `forceRetry` flag that the retry button can use in the ui to restart the cycle over.